diff --git a/common/tables.c b/common/tables.c
index c201b4b..d625126 100644
--- a/common/tables.c
+++ b/common/tables.c
@@ -475,6 +475,10 @@ static struct option dhcpv6_options[] = {
 	{ "clt-time", "L",			&dhcpv6_universe, 46, 1 },
 	{ "lq-relay-data", "6X",		&dhcpv6_universe, 47, 1 },
 	{ "lq-client-link", "6A",		&dhcpv6_universe, 48, 1 },
+
+	/* RFC 6334 Dual-Stack Lite AFTR-Name OPTIONS */
+	{ "aftr-name", "D",				&dhcpv6_universe,  64, 1 },
+
 	{ NULL, NULL, NULL, 0, 0 }
 };
 
diff --git a/includes/dhcp6.h b/includes/dhcp6.h
index df08337..67ae3d0 100644
--- a/includes/dhcp6.h
+++ b/includes/dhcp6.h
@@ -76,6 +76,8 @@
 #define D6O_LQ_RELAY_DATA			47 /* RFC5007 */
 #define D6O_LQ_CLIENT_LINK			48 /* RFC5007 */
 
+#define D6O_AFTR_NAME			64 /* RFC6334 */
+
 /* 
  * Status Codes, from RFC 3315 section 24.4, and RFC 3633, 5007.
  */
