diff -Nur a/SECboot/sec_boot.c b/SECboot/sec_boot.c
--- a/SECboot/sec_boot.c	2014-01-08 18:35:00.000000000 +0800
+++ b/SECboot/sec_boot.c	2014-01-08 17:32:10.000000000 +0800
@@ -19,16 +19,27 @@
 #include <exports.h>
 #include "sec_boot.h"
 
-#define LENGTH 				0x200000
 #define TMP_KERNEL_ADDR		0xA4000000
 #define REAL_KERNEL_ADDR	0xA0800000
 
 int sec_boot (int argc, char *argv[])
 {
+	uint32_t p0, p1, p2, p3;
+	uint8_t *ptr;
+	uint32_t size;
+		
 	printf("<%30s:%5d>: Decrypt kernel at %x\n", __func__, __LINE__, TMP_KERNEL_ADDR);
 	
-	arc_img_dec(TMP_KERNEL_ADDR, LENGTH, NULL);
-	my_memcpy(REAL_KERNEL_ADDR, TMP_KERNEL_ADDR, LENGTH);
+	ptr=TMP_KERNEL_ADDR;
+	p0=ptr[0];
+	p1=ptr[1];
+	p2=ptr[2];
+	p3=ptr[3];
+	size=p3<<24|p2<<16|p1<<8|p0;
+	printf("<%30s:%5d>: kernel size = %x\n", __func__, __LINE__, size);
+
+	arc_img_dec(TMP_KERNEL_ADDR+(sizeof(uint32_t)), size, NULL);
+	my_memcpy(REAL_KERNEL_ADDR, TMP_KERNEL_ADDR+(sizeof(uint32_t)), size);
 	return 0;
 }
 
