--- a/extensions/libipt_TRIGGER.c	2012-08-29 20:12:51.433280971 +0800
+++ b/extensions/libipt_TRIGGER.c	2012-08-29 20:37:21.673281003 +0800
@@ -202,6 +202,8 @@
 	}
 	printf(" ");
 
+	if(info->mport[0] > 0)
+	{
 	printf("match:%hu", info->mport[0]);
 	if (info->mport[1] > info->mport[0])
 		printf("-%hu", info->mport[1]);
@@ -210,6 +212,10 @@
 	printf("relate:");
 	for(i = 0; i < MAX_TRIG_RANGE; i ++)
 	{
+		if(info->rport[i][0] == 0)
+		{
+			break;
+		}
 		if(i > 0)
 		{
 			printf(",");
@@ -219,6 +225,7 @@
 			printf("-%hu", info->rport[i][1]);
 	}
 	printf(" ");
+	}
 }
 
 /* Saves the union ipt_targinfo in parsable form to stdout. */
@@ -251,6 +258,8 @@
 	}
 	printf(" ");
 	
+	if(info->mport[0] > 0)
+	{
 	printf("--trigger-match %hu", info->mport[0]);
 	if (info->mport[1] > info->mport[0])
 		printf("-%hu", info->mport[1]);
@@ -259,6 +268,10 @@
 	printf("--trigger-relate ");
 	for(i = 0; i < MAX_TRIG_RANGE; i ++)
 	{
+		if(info->rport[i][0] == 0)
+		{
+			break;
+		}
 		if(i > 0)
 		{
 			printf(",");
@@ -268,6 +281,7 @@
 			printf("-%hu", info->rport[i][1]);
 	}
 	printf(" ");
+	}
 }
 
 static
