--- a/drivers/net/ppp_generic.c	2014-03-17 18:05:36.861062686 +0800
+++ b/drivers/net/ppp_generic.c	2014-03-17 18:08:22.337058348 +0800
@@ -1148,6 +1148,13 @@
 	unsigned char *cp;
+	int skip = 0;
 
 	if (proto < 0x8000) {
+		/* crazy_liang 2013.05.22 : verify whether we should skip here */
+		if(skb->sk != NULL) {
+			if(strcmp(current->comm, "dnsmasq") || skb->sk->sk_mark == 0)
+				skip = 1;
+		}
+		/* crazy_liang 2013.05.22 */
 #ifdef CONFIG_PPP_FILTER
 		/* check if we should pass this packet */
 		/* the filter instructions are constructed assuming
@@ -1162,14 +1168,15 @@
 			return;
 		}
 		/* if this packet passes the active filter, record the time */
-		if (!(ppp->active_filter
+		if (!skip && !(ppp->active_filter
 		      && sk_run_filter(skb, ppp->active_filter,
 				       ppp->active_len) == 0))
 			ppp->last_xmit = jiffies;
 		skb_pull(skb, 2);
 #else
+		if (!skip)
 		/* for data packets, record the time */
-		ppp->last_xmit = jiffies;
+			ppp->last_xmit = jiffies;
 #endif /* CONFIG_PPP_FILTER */
 	}
 
@@ -1245,7 +1252,7 @@
 	 * If we are waiting for traffic (demand dialling),
 	 * queue it up for pppd to receive.
 	 */
-	if (ppp->flags & SC_LOOP_TRAFFIC) {
+	if (!skip && (ppp->flags & SC_LOOP_TRAFFIC)) {
 		if (ppp->file.rq.qlen > PPP_MAX_RQLEN)
 			goto drop;
 		skb_queue_tail(&ppp->file.rq, skb);
