diff --git a/pppoecd/auth.c b/pppoecd/auth.c
index a0721cc..7b1322e 100644
--- a/pppoecd/auth.c
+++ b/pppoecd/auth.c
@@ -545,8 +545,9 @@ auth_reset(unit)
 {
     lcp_options *ao = &lcp_allowoptions[0];
 
-    ao->neg_upap = !refuse_pap && (passwd[0] != 0);
-    ao->neg_chap = !refuse_chap	&& (passwd[0] != 0);
+	/* Zhijian 2015-12-14 Modify to allow empty password. */
+    ao->neg_upap = !refuse_pap /*&& (passwd[0] != 0)*/;
+    ao->neg_chap = !refuse_chap/*&& (passwd[0] != 0)*/;
 }
 
 /*
