Index: linux-3.5-rc2/include/linux/skbuff.h
===================================================================
--- linux-3.5-rc2.orig/include/linux/skbuff.h	2012-06-09 03:40:09.000000000 +0200
+++ linux-3.5-rc2/include/linux/skbuff.h	2012-06-13 13:20:13.273919848 +0200
@@ -1894,10 +1894,15 @@
 static inline int __skb_cow(struct sk_buff *skb, unsigned int headroom,
 			    int cloned)
 {
+	unsigned int alloc_headroom = headroom;
 	int delta = 0;
 
 	if (headroom > skb_headroom(skb))
-		delta = headroom - skb_headroom(skb);
+		alloc_headroom = NET_SKB_PAD;
+	if (headroom > skb_headroom(skb) ||
+	    (cloned && alloc_headroom > skb_headroom(skb))) {
+		delta = alloc_headroom - skb_headroom(skb);
+	}
 
 	if (delta || cloned)
 		return pskb_expand_head(skb, ALIGN(delta, NET_SKB_PAD), 0,
